### PR TITLE
#2986 Support embedding content from file

### PR DESCRIPTION
**Pull request by jmsktm**
```
$ cat template.yaml
name: "@{name}"
spec:
  description: 'This is just a multistep sample plugin'
  runner:
    image: ubuntu
    namespace: signadot
    podTemplateOverlay: "@{embed:pod-template-overlay.yaml}"
  create:
  - name: provision
    inputs:
    - name: dbname
      valueFromSandbox: true
    outputs:
    - name: dbhost
      from: "/tmp/dbhost"
    - name: dbpass
      from: "/tmp/dbpass"
    script: "@{embed:provision.sh}"
  - name: setup
    inputs:
    - name: dbhost
      valueFromStep:
        name: provision
        output: dbhost
    - name: dbpass
      valueFromStep:
        name: provision
        output: dbpass
    - name: dbname
      valueFromSandbox: true
    outputs:
    - name: user-id
      from: "/tmp/user-id"
    script: "@{embed:setup.sh}"
  delete:
  - name: deprovision
    inputs:
    - name: dbhost
      valueFromStep:
        name: provision
        output: dbhost
    - name: dbpass
      valueFromStep:
        name: provision
        output: dbpass
    - name: dbname
      valueFromSandbox: true
    outputs:
    - name: op-id
      from: "/tmp/op-id"
    script: "@{embed:deprovision.sh}"
  - name: op-releaser
    inputs:
    - name: op-id
      valueFromStep:
        name: deprovision
        output: op-id
    - name: user-id
      valueFromStep:
        name: setup
        output: user-id
    script: "@{embed:op-releaser.sh}"
```

```
$ cat pod-template-overlay.yaml
metadata:
  namespace: signadot
spec:
  serviceAccountName: sd-mariadb
```

```
$ cat provision.sh
#!/bin/bash
echo "Provisioning DB ${DBNAME}"
sleep 10
echo "passwd123" > /tmp/dbpass
echo "mariabdb.svc" > /tmp/dbhost
```

```
$ cat setup.sh
#!/bin/bash
echo "Seeding DB ${DBNAME}"
echo "DB_PASSWORD = $DB_PASSWORD"
echo "DB_HOST = $DB_HOST"
sleep 10
echo "1" > /tmp/user-id
```

```
$ cat deprovision.sh
#!/bin/bash
echo "Deprovision DB $DBNAME"
echo "DB_PASSWORD = $DB_PASSWORD"
echo "DB_HOST = $DB_HOST"
sleep 10
echo "123456" > /tmp/op-id
```

```
$ cat op-releaser.sh
#!/bin/bash
echo "Op Releaser"
echo "OP_ID = $OP_ID"
echo "USER_ID = $USER_ID"
sleep 10
```

```
$ ./signadot --config ~/.signadot/localhost.yaml resourceplugin apply -f template.yaml --set name=plugin-1
Created resource plugin with name "plugin-1"
```

```
$ ./signadot --config ~/.signadot/localhost.yaml resourceplugin get plugin-1
Name:      plugin-1
Created:   Wed, 15 Mar 2023 20:57:08 PDT (35 seconds ago)
Status:    0 resources
```

```
$ ./signadot --config ~/.signadot/localhost.yaml resourceplugin get plugin-1 -o yaml
createdAt: "2023-03-16T03:57:08.469371Z"
name: plugin-1
spec:
  create:
  - inputs:
    - name: dbname
      valueFromSandbox: true
    name: provision
    outputs:
    - from: /tmp/dbhost
      name: dbhost
    - from: /tmp/dbpass
      name: dbpass
    script: |
      #!/bin/bash
      echo "Provisioning DB ${DBNAME}"
      sleep 10
      echo "passwd123" > /tmp/dbpass
      echo "mariabdb.svc" > /tmp/dbhost
  - inputs:
    - name: dbhost
      valueFromStep:
        name: provision
        output: dbhost
    - name: dbpass
      valueFromStep:
        name: provision
        output: dbpass
    - name: dbname
      valueFromSandbox: true
    name: setup
    outputs:
    - from: /tmp/user-id
      name: user-id
    script: |
      #!/bin/bash
      echo "Seeding DB ${DBNAME}"
      echo "DB_PASSWORD = $DB_PASSWORD"
      echo "DB_HOST = $DB_HOST"
      sleep 10
      echo "1" > /tmp/user-id
  delete:
  - inputs:
    - name: dbhost
      valueFromStep:
        name: provision
        output: dbhost
    - name: dbpass
      valueFromStep:
        name: provision
        output: dbpass
    - name: dbname
      valueFromSandbox: true
    name: deprovision
    outputs:
    - from: /tmp/op-id
      name: op-id
    script: |
      #!/bin/bash
      echo "Deprovision DB $DBNAME"
      echo "DB_PASSWORD = $DB_PASSWORD"
      echo "DB_HOST = $DB_HOST"
      sleep 10
      echo "123456" > /tmp/op-id
  - inputs:
    - name: op-id
      valueFromStep:
        name: deprovision
        output: op-id
    - name: user-id
      valueFromStep:
        name: setup
        output: user-id
    name: op-releaser
    outputs: null
    script: |
      #!/bin/bash
      echo "Op Releaser"
      echo "OP_ID = $OP_ID"
      echo "USER_ID = $USER_ID"
      sleep 10
  description: This is just a multistep sample plugin
  runner:
    image: ubuntu
    namespace: signadot
    podTemplateOverlay: |
      metadata:
        namespace: signadot
      spec:
        serviceAccountName: sd-mariadb
status:
  resources: []
updatedAt: "2023-03-16T03:57:08.469371Z"
```

### CODE REVIEW

**Review by scott-cotton — COMMENTED** [review]
Great to see this working and the tests and comments.

Food for thought: Is this forward compatible with being able to embed actual yaml/json?  I guess we could have an `embed-yaml` operator, but the idea of specifying the type at the level of operator doesn't seem well-placed to me.  

I'd suggest we think about if this can be extended to embedding yaml without introducing awkwardness.  Handling everything structured and typed as a string is like `tcl`: it has its place but there are many contexts in which the type is much more helpful.  I'd like to avoid making it hard to embed arbitrary yaml just for expediency if at all possible.

**Comment by daniel-de-vera** [comment]
> @scott-cotton , @daniel-de-vera : It only looks for the files in the current path. If we want to be able to support embedding file content from user provided path, I'll make that change separately.

Yes, I was kind of counting on that. If you see the new plugins I'm working on, I'm assuming I can embed files from a different directory:

https://github.com/signadot/plugins/tree/resources-v2

File: internal/config/flag_types.go [hunk 1]
```diff
@@ -36,7 +36,44 @@ func (o *OutputFormat) Type() string {
...
+			// Reasoning for grouping within the regex pattern:
+		  // 1. The complete pattern matches `@{...}`
+		  // 2. The content of `@{...}` matches `(optional-part)?(value-match)`
+		  // 3. The optional part contains an operation followed by a colon. Eg. `(operation):`
+		  // 4. We only support 1 operation (embed) at the moment. But to support more in the future, it is
+		  //    grouped as well so that the supported operations can be marked as one of. Eg. (embed|import)
+		  // 5. For the value (or variable?), we'll only support alphanumeric characters and dots.
+		  //    So, no support for path in case of embedding at the moment.
+	*/
+	VarRefRx = regexp.MustCompile(`\@\{(((embed):)?([a-zA-Z][a-zA-Z0-9_.-]*))\}`)
```
[the remark this answers is not in the document]
Comment (jmsktm) [reply internal/config/flag_types.go]: I didn't use the exact regex, but have made it more relaxed for both directive operation key and operand, and have allowed spaces surrounding them. PTAL.

**Comment by jmsktm** [comment]
@scott-cotton , @daniel-de-vera : I have made some changes to the parser and the regex pattern to make the components of the placeholder strings (`${...}`) more relaxed. But I'm yet to address the two things below:

- ensure forward compatibility to support structured data (JSON/YAML) in addition to string. I plan to look into that prior to merging in this PR.
- support for path: I will look into that separately after merging this in.

**Comment by daniel-de-vera** [comment]
@jmsktm, one issue that I found is that the include path is relative to the place where you are running the CLI, instead of being from yaml path that you are applying, e.g.:

```
$ go run ./cmd/signadot resourceplugin apply -f /home/ddv/git/signadot/plugins/mariadb/plugin.yaml 
Error: error reading from file provision.sh: open provision.sh: no such file or directory
exit status 1
```

Where:

```yaml
  create:
  - name: provision
    inputs:
    - name: dbname
      valueFromSandbox: true
      to:
        env: DBNAME

    outputs:
    - name: host
      from: /tmp/host
    - name: port
      from: /tmp/port
    - name: root-password
      from: /tmp/root-password

    script: "@{embed:provision.sh}"
```

And:

```
$ ls -l /home/ddv/git/signadot/plugins/mariadb/
total 24
-rw-rw-r-- 1 ddv ddv  404 mar 16 10:44 deprovision.sh
-rw-rw-r-- 1 ddv ddv  811 mar 16 10:46 plugin.yaml
-rw-rw-r-- 1 ddv ddv 1151 mar 16 10:44 provision.sh
```

For now, I can solve it by building the CLI, and cd into the plugin directory, but I think this is something important to consider when we implement paths, because the relative path should be against the location of the current file, e.g.:

```yaml
  delete:
    - name: deprovision
      script: "@{embed:./plugin/deprovision.sh}"
```
```
$ find
./plugin.yaml
./plugin/deprovision.sh
```

**Review by daniel-de-vera — APPROVED** [review]
I created all plugins from https://github.com/signadot/plugins/tree/resources-v2 (moving files to avoid the `./plugin/ `sub-directories) and some matching sandboxes, and everything worked fine.
I also successfully tested regular templates  (e.g.: `@{route-image-tag}`).
So, LGTM.

**Comment by scott-cotton** [comment]
> * support for path: I will look into that separately after merging this in.

I looked at this more closely and it seems to me we should refactor the parsing code so it doesn't work by regexp.   Let me know if you'd like to discuss or help.

File: internal/config/flag_types.go [hunk 2]
```diff
@@ -36,7 +36,31 @@ func (o *OutputFormat) Type() string {
...
+		//  4 - 26: @{^embed^:^file.yaml^} (directive string)
+		//  7 - 12: embed                  (group #1: operation)
+		// 15 - 24: file.yaml              (group #2: operand)
+
+		// [31 39 -1 -1 34 37]             # plain variable substitution
+		// 31 - 39: @{^dev^}               (placeholder string)
+		// -1 - -1: NO_MATCH               (group #1: no operation to match)
+		// 34 - 37: dev                    (group #2: variable name)
+	*/
+	VarRefRx = regexp.MustCompile(`@{\s*(?:([a-zA-Z][a-zA-Z0-9_.-]*)\s*:)?\s*([a-zA-Z0-9_.\\/-]*)\s*}`)
```
Comment (scott-cotton) [inline internal/config/flag_types.go:63]: I think we can move this to the util package unexported and delete the old parsing code.

**Comment by jmsktm** [comment]
> For now, I can solve it by building the CLI, and cd into the plugin directory, but I think this is something important to consider when we implement paths, because the relative path should be against the location of the current file, e.g.:

@daniel-de-vera : As of now, yes. But with the change I'll be making today to support paths, it will be relative to the location of the template.

> I looked at this more closely and it seems to me we should refactor the parsing code so it doesn't work by regexp. Let me know if you'd like to discuss or help.

Sure, I'm open to that. But I suppose that refactoring is for later, and not blocking for this PR?

Btw, I'm also considering refactoring the templating code to be better structured before building more on top of that. But I am planning that effort for later.

**Comment by jmsktm** [comment]
@scott-cotton : I have made the change to support embedding YAML using `embed[yaml]`.

**Pull request merged**